### PR TITLE
fix(language): canonicalize bridge aliases before parser load

### LINKED ISSUE #691: Bridge language aliases depend on parser availability
## Problem

Injection language canonicalization uses `InjectionResolver::resolve_language`, which delegates to `LanguageCoordinator::detect_language_hot`. That detection only returns a canonical candidate when a parser is already available (directly or through `base`).

For a `py` fence when the Python parser is not loaded yet, resolution therefore falls back to raw `py`. Eager bridge lookup performs exact matching against `languageServers[*].languages`, so a normal `languages = ["python"]` server is skipped even though the token alias is known. The same applies to first-line detection such as a Python shebang when its parser is unavailable.

The background injection install check later normalizes `py` for parser installation, but eager bridge selection has already received only the raw language and content-free language set.

## Evidence

- `src/language/injection/discovery.rs`: `resolve_language` falls back to the raw identifier when `detect_language_hot` returns `None`.
- `src/language/coordinator.rs`: `try_with_base_fallback` returns a candidate only when `has_parser_available` succeeds.
- `src/lsp/bridge/coordinator.rs`: bridge configs match `injection_language` exactly.
- `src/lsp/lsp_impl/coordinator/injection.rs`: auto-install normalizes the language later, after eager discovery.

## Expected behavior

Bridge language selection should preserve a canonical token/first-line/base candidate independently of whether the Tree-sitter parser is already loaded, so `py` can select a `python` bridge during the initial eager lifecycle pass.

## Context

Discovered while reviewing #687. That issue is about raw-vs-normalized divergence between eager lifecycle and interactive request paths; sharing the same resolver fixes that divergence. Parser-availability-dependent canonicalization is a separate policy problem common to both paths.

### PR DESCRIPTION
**Pull request by atusy**
## Stack dependency

This PR is intentionally stacked on #692 (`fix/issue-687-canonical-injection-language`). #692 establishes the shared `ResolvedInjection` path used by eager and interactive bridge dispatch; this PR changes that shared resolver policy. Review only the commits/diff relative to the #692 branch. Merge #692 first, then retarget this PR to `main`.

## Summary

- canonicalize bridge injection tokens, configured bases, and first-line candidates without consulting parser availability
- keep bridge language keys and virtual-document URIs stable before and after parser installation
- preserve configured base priority over generic syntect aliases
- document the separate bridge-canonicalization and parser-resolution contracts in the language-detection ADR

## Regression coverage

- `py` resolves to `python` before and after registering a parser under raw alias `py`
- explicit `py` beats a conflicting loaded shebang language
- configured `py.base` overrides generic token normalization without a parser
- unknown identifier uses Python shebang without a loaded parser
- configured `plaintext.base` is honored for bridge identity

## Validation

- `TREE_SITTER_GRAMMARS=/Users/atusy/ghq/github.com/atusy/kakehashi___main/deps/tree-sitter cargo test --lib -q` — 2799 passed, 2 ignored
- focused alias/base/first-line regressions
- `make check`

Closes #691

## Summary by CodeRabbit

* **Documentation**
  * Updated migration guidance from `aliases` to `base`, including language inheritance and canonicalization behavior.
  * Reworked language detection fallback documentation to reflect parser-availability-aware selection while keeping stable canonical identity.

* **Bug Fixes**
  * Improved injected-language canonicalization with clearer precedence across configured `base`, explicit tokens, and heuristics.
  * Tightened `plaintext` handling so it resolves only via eligible configured mappings; otherwise it stays unresolved.
  * Prevented injected-language auto-install/load when injection parsing is disabled.

* **Tests**
  * Expanded injected-language canonicalization and `plaintext` edge-case coverage.

### CODE REVIEW

File: src/language/coordinator.rs [hunk 1]
```diff
@@ -1114,21 +1112,26 @@ impl LanguageCoordinator {
...
+            return base;
+        }
+        if let Some(candidate) = super::heuristic::detect_from_token(identifier) {
+            return self.resolve_base(&candidate).unwrap_or(candidate);
+        }
+        if let Some(candidate) = super::heuristic::detect_from_first_line(content) {
+            return self.resolve_base(&candidate).unwrap_or(candidate);
+        }
+        identifier.to_string()
     }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs:1254]: Refuted as out of scope/inapplicable. canonical_injection_language receives one already-discovered injection region; it performs no document scan and has no cursor parameter. Cursor-local region discovery is handled by resolve_at_byte_offset before this function, while resolve_all intentionally serves whole-document methods. Changing discovery here cannot improve that scaling.

File: src/language/injection/discovery.rs [hunk 2]
```diff
@@ -524,14 +524,7 @@ impl InjectionResolver {
...
-        coordinator
-            .detect_language_hot(
-                raw_identifier, // Use identifier as path for token extraction
-                content,
-                Some(raw_identifier), // Explicit token
-                Some(raw_identifier), // Also try as languageId
-            )
-            .unwrap_or_else(|| raw_identifier.to_string())
+        coordinator.canonical_injection_language(raw_identifier, content)
     }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/injection/discovery.rs:525]: Fixed in 2c5817fed. resolve_language now documents parser-independent bridge/URI canonicalization and its explicit raw-key fallback. make check passes.

File: src/language/coordinator.rs [hunk 3]
```diff
@@ -1101,9 +1101,7 @@ impl LanguageCoordinator {
     /// (1) LSP `languageId` (if not `"plaintext"`); (2) heuristics — explicit
     /// token (`"py"`, `"js"`), path token via `extract_token_from_path`, then
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in 2c5817fed. detect_language is now documented as the parser-aware host-document chain, distinct from bridge canonicalization. make check passes.

File: src/language/coordinator.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs:1254]: Refuted after control-flow audit. resolve_base executes at most twice, not three times: a token candidate returns immediately, and first-line detection runs only when token detection returned None. Holding one RwLock guard across syntect token/first-line heuristics would lengthen writer blocking on this typing hot path. Without measured contention, the current tightly scoped reads are preferable.

[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs:1254]: Refuted again; the control flow is unchanged. resolve_base runs at most twice, never three times: token success returns immediately, and first-line detection only runs when token detection returned None. The suggested single guard would stay held across syntect token/first-line heuristic work, increasing writer blocking versus the current tightly scoped reads. There is no measured contention supporting that trade-off; make check and the full local suite pass.

File: docs/architecture-decisions/language-detection-fallback-chain.md [hunk 4]
```diff
@@ -52,77 +59,78 @@ Each detection method follows the **detect → alias resolution → availability
...
     if effective_token exists:
         // Try syntect normalization (py → python, Makefile → make)
         if syntect recognizes token:
-            if try_with_alias_fallback(normalized) succeeds:
+            if try_with_base_fallback(normalized) succeeds:
                 return result
-        // Try raw token as alias (handles jsx, tsx that syntect doesn't know)
-        if try_with_alias_fallback(raw_token) succeeds:
+        // Try raw token with base fallback (handles jsx, tsx)
+        if try_with_base_fallback(raw_token) succeeds:
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md]: Fixed in b7fa7965f. The direct/base fallback now also uses effective_token consistently.

File: docs/architecture-decisions/language-detection-fallback-chain.md [hunk 5]
```diff
@@ -52,77 +59,78 @@ Each detection method follows the **detect → alias resolution → availability
...
     if language_id exists and != "plaintext":
-        if try_with_alias_fallback(language_id) succeeds:
+        if try_with_base_fallback(language_id) succeeds:
             return result
 
     // 2. Token-based detection
     effective_token = token OR extract_token_from_path(path)
     if effective_token exists:
         // Try syntect normalization (py → python, Makefile → make)
         if syntect recognizes token:
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md]: Fixed in b7fa7965f. The syntect step now names the defined effective_token variable.

File: src/language/coordinator.rs [hunk 6]
```diff
@@ -1114,21 +1112,26 @@ impl LanguageCoordinator {
...
-        language_id: Option<&str>,
-    ) -> Option<String> {
-        self.detect_language_logged(path, content, token, language_id, log::Level::Trace)
+    /// Candidate selection deliberately does not inspect parser state: eager
+    /// bridge selection and virtual URIs must stay stable before and after a
+    /// parser loads. A configured base for the explicit identifier takes
+    /// precedence over generic token aliases, followed by first-line fallback.
+    /// Consequently, registering a parser under an alias such as `py` or `js`
+    /// does not change bridge keys/URIs: bridge configuration uses the canonical
+    /// `python`/`javascript` name unless that alias has an explicit base.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in ab283ffb1. The comment now says syntect token normalization and non-canonical parser key, avoiding ambiguity with the deprecated aliases config field. make check passes.

File: src/language/injection/discovery.rs [hunk 7]
```diff
@@ -665,6 +659,95 @@ mod tests {
...
+                &coordinator,
+                "py",
+                "#!/usr/bin/env lua\nprint('hello')",
+            ),
+            "python"
+        );
+    }
+
+    #[test]
+    fn configured_base_beats_generic_token_alias_without_parser() {
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/injection/discovery.rs]: Fixed in bbb9b07e4. Renamed the regression to configured_base_beats_syntect_token_normalization_without_parser; focused test and make check pass.

File: docs/architecture-decisions/language-detection-fallback-chain.md [hunk 8]
```diff
@@ -52,77 +59,78 @@ Each detection method follows the **detect → alias resolution → availability
...
 
     // 2. Token-based detection
     effective_token = token OR extract_token_from_path(path)
     if effective_token exists:
         // Try syntect normalization (py → python, Makefile → make)
-        if syntect recognizes token:
-            if try_with_alias_fallback(normalized) succeeds:
+        if syntect recognizes effective_token:
+            if try_with_base_fallback(normalized) succeeds:
                 return result
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md:112]: Fixed in bbb9b07e4. The pseudocode now explicitly assigns normalized = syntect_detect(effective_token). make check passes.

File: docs/architecture-decisions/language-detection-fallback-chain.md [hunk 9]
```diff
@@ -52,77 +59,78 @@ Each detection method follows the **detect → alias resolution → availability
...
-        if try_with_alias_fallback(raw_token) succeeds:
+        // Try raw token with base fallback (handles jsx, tsx)
+        if try_with_base_fallback(effective_token) succeeds:
             return result
 
     // 3. First-line detection (shebang, mode line)
     if syntect detects from first line:
-        if try_with_alias_fallback(detected) succeeds:
+        if try_with_base_fallback(detected) succeeds:
             return result
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md:120]: Fixed in 2f9d7344f. The pseudocode now assigns detected = syntect_detect_first_line(content) before base fallback.

[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md:120]: Correction: fixed in 0048212ef (the prior reply contained a mistyped hash). The pseudocode now assigns the first-line candidate explicitly.

File: src/language/injection/discovery.rs [hunk 10]
```diff
@@ -512,26 +512,20 @@ impl InjectionResolver {
...
+    /// Derive a parser-independent canonical injection language for bridge
+    /// selection and stable virtual-document identity.
     ///
     /// This normalizes raw fence identifiers (e.g., "py") to canonical language names
     /// (e.g., "python") that match bridge server configurations.
     ///
-    /// Falls back to the raw identifier if no resolution is found, allowing the
-    /// bridge lookup to fail gracefully with a clear error message.
+    /// Falls back to the raw identifier when no configured or heuristic
+    /// canonical candidate exists; bridge matching then uses that explicit key.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/injection/discovery.rs]: Fixed in e3390441d. resolve_language now describes the stable bridge key as coming from an explicit configured base or heuristic normalization.

File: docs/architecture-decisions/language-detection-fallback-chain.md [hunk 11]
```diff
@@ -134,12 +142,28 @@ Document (markdown) ──parse──▶ AST ──injection query──▶ "py"
...
-3. **Try config-based alias**: If syntect doesn't recognize it, check user-configured aliases
-4. **Skip if unavailable**: If no parser matches, the region is skipped
+For example, a Markdown code fence with ` ```py ` provides the identifier
+`"py"`. Two related operations intentionally have different acceptance rules:
+
+- **Bridge canonicalization** produces a stable language key even when no
+  parser exists yet. It checks an explicit configured base, then syntect token
+  normalization (and that candidate's base), then first-line detection (and
+  that candidate's base), and finally the raw identifier. Thus `py` remains
+  `python` before and after parser installation.
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md]: Fixed in e3390441d. The ADR now qualifies py-to-python stability with the absence of an explicit py base override.

File: src/language/coordinator.rs [hunk 12]
```diff
@@ -1225,21 +1223,26 @@ impl LanguageCoordinator {
...
+            return base;
+        }
+        if let Some(candidate) = super::heuristic::detect_from_token(identifier) {
+            return self.resolve_base(&candidate).unwrap_or(candidate);
+        }
+        if let Some(candidate) = super::heuristic::detect_from_first_line(content) {
+            return self.resolve_base(&candidate).unwrap_or(candidate);
+        }
+        identifier.to_string()
     }
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs:1254]: Fixed in 177b85916. Eligible plaintext base mappings still win, but unconfigured explicit plaintext now returns plaintext before token/first-line heuristics. Added a shebang regression; focused tests and make check pass.

File: src/language/coordinator.rs [hunk 13]
```diff
@@ -1225,21 +1223,26 @@ impl LanguageCoordinator {
...
-        content: &str,
-        token: Option<&str>,
-        language_id: Option<&str>,
-    ) -> Option<String> {
-        self.detect_language_logged(path, content, token, language_id, log::Level::Trace)
+    /// Candidate selection deliberately does not inspect parser state: eager
+    /// bridge selection and virtual URIs must stay stable before and after a
+    /// parser loads. A configured base for the explicit identifier takes
+    /// precedence over syntect token normalization, followed by first-line
+    /// fallback. Consequently, registering a parser under a non-canonical key
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in 0a3ce02fb. canonical_injection_language and resolve_base now describe eligible registered base mappings and explicitly note that distinct-parser declarations are not registered.

File: src/language/coordinator.rs [hunk 14]
```diff
@@ -1381,10 +1391,20 @@ impl LanguageCoordinator {
             content.len()
         );
 
-        // 1. Try direct identifier first (skip "plaintext")
-        if identifier != "plaintext"
-            && let Some(found) = self.try_load_with_base(identifier)
-        {
+        // 1. Try the identifier first. Explicit plaintext remains featureless
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs:1397]: Fixed in 4ff042a43. The rustdoc now explicitly documents plaintext as the exception: only an eligible configured base is loaded, otherwise it returns None without token or first-line heuristics.

File: src/language/coordinator.rs [hunk 15]
```diff
@@ -1225,21 +1225,31 @@ impl LanguageCoordinator {
...
-        self.detect_language_logged(path, content, token, language_id, log::Level::Trace)
+    /// Candidate selection deliberately does not inspect parser state: eager
+    /// bridge selection and virtual URIs must stay stable before and after a
+    /// parser loads. An eligible configured base mapping for the explicit
+    /// identifier takes precedence. An otherwise unconfigured explicit
+    /// `plaintext` remains `plaintext`; other identifiers proceed through syntect
+    /// token normalization and then first-line fallback. Consequently,
+    /// registering a parser under a non-canonical key such as `py` or `js` does
+    /// not change bridge keys/URIs: bridge configuration uses
+    /// `python`/`javascript` unless that key has a registered base mapping.
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in 2de080438. The rustdoc now explicitly names the raw identifier as the base-mapping override and gives py.base as the example.

File: src/language/coordinator.rs [hunk 16]
```diff
@@ -1931,6 +1953,42 @@ mod tests {
         assert!(load_result.success);
     }
 
+    #[test]
+    fn test_injection_plaintext_uses_eligible_config_base() {
+        let coordinator = LanguageCoordinator::new();
+        coordinator
+            .language_registry_for_parallel()
+            .register("python".to_string(), tree_sitter_rust::LANGUAGE.into());
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in 2de080438. The plaintext-base test now registers tree_sitter_python under the python key.

File: src/language/coordinator.rs [hunk 17]
```diff
@@ -1931,6 +1953,42 @@ mod tests {
...
+        assert_eq!(resolved, "python");
+        assert!(load_result.success);
+    }
+
+    #[test]
+    fn test_injection_plaintext_without_base_stays_featureless() {
+        let coordinator = LanguageCoordinator::new();
+        coordinator
+            .language_registry_for_parallel()
+            .register("python".to_string(), tree_sitter_rust::LANGUAGE.into());
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in 2de080438. The featureless-plaintext fixture now also uses the matching Python grammar.

File: src/language/coordinator.rs [hunk 18]
```diff
@@ -1931,6 +1954,42 @@ mod tests {
         assert!(load_result.success);
     }
 
+    #[test]
+    fn test_injection_plaintext_uses_eligible_config_base() {
+        let coordinator = LanguageCoordinator::new();
+        coordinator
+            .language_registry_for_parallel()
+            .register("python".to_string(), tree_sitter_rust::LANGUAGE.into());
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Already fixed in 2de080438: this fixture now registers tree_sitter_python::LANGUAGE under the python key; focused plaintext tests, check, and clippy pass.

File: src/language/coordinator.rs [hunk 19]
```diff
@@ -1931,6 +1954,42 @@ mod tests {
...
+        assert_eq!(resolved, "python");
+        assert!(load_result.success);
+    }
+
+    #[test]
+    fn test_injection_plaintext_without_base_stays_featureless() {
+        let coordinator = LanguageCoordinator::new();
+        coordinator
+            .language_registry_for_parallel()
+            .register("python".to_string(), tree_sitter_rust::LANGUAGE.into());
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Already fixed in 2de080438: the second plaintext fixture also uses tree_sitter_python::LANGUAGE. This is a re-anchored duplicate of the resolved current-head finding.

[the remark this answers is not in the document]
Comment (atusy) [reply src/language/coordinator.rs]: Fixed in 6ab3a1de9. Both remaining python-key fixtures now register tree_sitter_python, including the first-line detection sibling. Focused tests, check, and clippy pass.

File: docs/architecture-decisions/language-detection-fallback-chain.md [hunk 20]
```diff
@@ -23,19 +26,23 @@ This applies to both document-level language detection and injected language res
...
                                                 →  If no: continue
-2. Token detection →  syntect     →  Try alias  →  If available: use it
-                  →  raw token   →  Try alias  →  If available: use it
+2. Token detection →  syntect     →  Try base   →  If available: use it
+                  →  raw token   →  Try base   →  If available: use it
                                                 →  If no: continue
-3. First line      →  Try direct  →  Try alias  →  If available: use it
+3. First line      →  Try direct  →  Try base   →  If available: use it
                                                 →  If no: return None
 ```
```
[the remark this answers is not in the document]
Comment (atusy) [reply docs/architecture-decisions/language-detection-fallback-chain.md]: Fixed in 08cb0ba76. All three previously unlabeled diagram/pseudocode fences now use text; cargo fmt and git diff checks pass.

File: src/lsp/lsp_impl/coordinator/injection.rs [hunk 21]
```diff
@@ -402,9 +405,21 @@ impl InjectionCoordinator {
     }
 }
 
+fn parser_enabled_injection_language(language: &str) -> bool {
+    // Explicit, unconfigured plaintext is deliberately featureless. A configured
+    // plaintext base is canonicalized to that base before reaching this loop.
+    language != "plaintext"
+}
```
[the remark this answers is not in the document]
Comment (atusy) [reply src/lsp/lsp_impl/coordinator/injection.rs:412]: Fixed in 70df26189. The comment now states that only an eligible configured plaintext base is canonicalized before this guard. Focused test, check, and clippy pass.
